pkg/instancetype/compatibility.go

```diff
@@ -1,40 +1,159 @@
+		return nil
+	case *instancetypev1alpha2.VirtualMachineInstancetype:
+		dest := &instancetypev1beta1.VirtualMachineInstancetype{}
+		if err := instancetypev1alpha2.Convert_v1alpha2_VirtualMachineInstancetype_To_v1beta1_VirtualMachineInstancetype(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha2.VirtualMachineClusterInstancetype:
+		dest := &instancetypev1beta1.VirtualMachineClusterInstancetype{}
+		if err := instancetypev1alpha2.Convert_v1alpha2_VirtualMachineClusterInstancetype_To_v1beta1_VirtualMachineClusterInstancetype(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha2.VirtualMachinePreference:
+		dest := &instancetypev1beta1.VirtualMachinePreference{}
+		if err := instancetypev1alpha2.Convert_v1alpha2_VirtualMachinePreference_To_v1beta1_VirtualMachinePreference(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha2.VirtualMachineClusterPreference:
+		dest := &instancetypev1beta1.VirtualMachineClusterPreference{}
+		if err := instancetypev1alpha2.Convert_v1alpha2_VirtualMachineClusterPreference_To_v1beta1_VirtualMachineClusterPreference(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha1.VirtualMachineInstancetype:
+		dest := &instancetypev1beta1.VirtualMachineInstancetype{}
+		if err := instancetypev1alpha1.Convert_v1alpha1_VirtualMachineInstancetype_To_v1beta1_VirtualMachineInstancetype(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha1.VirtualMachineClusterInstancetype:
+		dest := &instancetypev1beta1.VirtualMachineClusterInstancetype{}
+		if err := instancetypev1alpha1.Convert_v1alpha1_VirtualMachineClusterInstancetype_To_v1beta1_VirtualMachineClusterInstancetype(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha1.VirtualMachinePreference:
+		dest := &instancetypev1beta1.VirtualMachinePreference{}
+		if err := instancetypev1alpha1.Convert_v1alpha1_VirtualMachinePreference_To_v1beta1_VirtualMachinePreference(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	case *instancetypev1alpha1.VirtualMachineClusterPreference:
+		dest := &instancetypev1beta1.VirtualMachineClusterPreference{}
+		if err := instancetypev1alpha1.Convert_v1alpha1_VirtualMachineClusterPreference_To_v1beta1_VirtualMachineClusterPreference(obj, dest, nil); err != nil {
+			return err
+		}
+		revision.Data.Object = dest
+	default:
+		return fmt.Errorf("unexpected type in ControllerRevision: %T", obj)
+	}
+	return nil
+}
+
+func decodeSpecRevision(data []byte, isPreference bool) (runtime.Object, error) {
```

Comment:
nit: I wonder isPreference bool can be dropped in favor of two separate methods. IMHO it will be more readable

Reply:
Yup valid however this code is going to be dropped with the removal of `v1alpha1` and `decodeVirtualMachinePreferenceSpecRevision` support so I'd rather hold off refactoring until then. At that point we should be able to drop `isPreference` from the entire call stack above this and rely on `runtime.Decode` as we do with the later `ControllerRevisions`.